File: rb/morpion.rb
require 'set'

class Morpion

  H = HORIZONTAL = [1, 0]
  V = VERTICAL = [0, 1]
  D = DESCENDING = [1, 1]
  A = ASCENDING = [1, -1]
  
  def initialize(line_length, disjoint = 0)
    @l, @d = line_length, disjoint
    @data = {}
    @bound = Set.new
  end

  def mark(x, y, line = nil)
    @data[[x, y]] = {} # nil until played

    # extends the boundary to all open points around the mark
    @bound.delete [x, y]
    draw_bound x, y

    draw_line *line if line

    self
  end

  def unmark(x, y, line)
    clear_line *line
    @data.delete [x, y]
    clear_bound x, y
    
    # compensate for lost boundaries from other marks
    (x-2..x+2).each do |bx|
      (y-2..y+2).each do |by|
        draw_bound bx, by if @data[[bx, by]]
      end
    end
  end

  # same as mark, but raises an error on an invalid move
  def mark_safe(x, y, line)
    if valid_moves.include?([x, y, line])
      mark x, y, line
    else
      raise 'invalid move'
    end
  end

  def mark_all(marks)
    marks.each{|m| mark *m}
    self
  end

  def valid_moves
    @bound.inject([]) do |moves, (bx, by)|
      moves.concat lines(bx, by).map{|line| [bx, by, line]}
    end
  end

  def to_s(bound = false)
    b = bound ? 1 : 0
    (y_min-b..y_max+b).map do |y|
      (x_min-b..x_max+b).map do |x|
        if @data[[x, y]]
          'X'
        elsif bound && @bound.include?([x, y])
          '-'
        else
          ' '
        end
      end.join ''
    end.join "\n"
  end

  def x_min
    @data.keys.map(&:first).min
  end

  def x_max
    @data.keys.map(&:first).max
  end

  def y_min
    @data.keys.map(&:last).min
  end

  def y_max
    @data.keys.map(&:last).max
  end

  def id
    @data.hash
  end

  private

  def draw_bound(x, y)
    [x-1,x,x+1].product([y-1,y,y+1]).each do |(bx, by)|
      @bound << [bx, by] unless @data[[bx, by]]
    end
  end

  def clear_bound(x, y)
    [x-1,x,x+1].product([y-1,y,y+1]).each do |(bx, by)|
      @bound.delete [bx, by]
    end
  end

  def draw_line(x, y, dx, dy)
    (0..@l).each do |i|
      xi, yi = x+i*dx, y+i*dy
      @data[[xi, yi]][[dx, dy]] ||= Set.new
      @data[[xi, yi]][[dx, dy]] << [x, y]
    end
  end

  def clear_line(x, y, dx, dy)
    (0..@l).each do |i|
      xi, yi = x+i*dx, y+i*dy
      @data[[xi, yi]][[dx, dy]].delete [x, y]
      @data[[xi, yi]].delete [dx, dy] if @data[[xi, yi]][[dx, dy]].empty?
    end
  end

  # returns possible lines for a move made at _open_ spot x,y
  def lines(x, y)
    lines = []

    [H, V, D, A].each do |d|
      dx, dy = d

      # count marks ahead of, and behind, the current point
      behind, ahead = [-1, 1].map do |dir|
        adj = 0
        continuous = true
        (1..@l+@d).each do |i|
          if p = @data[[x+i*dx*dir, y+i*dy*dir]]
            adj += 1 if continuous
            if p[d]
              adj = [i-@d, adj].min
              break
            end
          else
            continuous = false
          end
        end
        [@l, adj].min
      end

      (-behind..ahead-@l).each do |i|
        lines << [x+i*dx, y+i*dy, dx, dy]
      end
    end

    lines
  end

end

File: js/v2/morpion.js
var Morpion = function(len, dis) {
  this.len = len;
  this.dis = dis;
  this.data = {};
  this.bound = {};
};


Morpion.ORIENTATIONS = [
  [1, 0],  // horiz
  [0, 1],  // vert
  [1, -1], // asec
  [1, 1]   // desc
];

Morpion.prototype = {

  mark: function(x, y, line) {
    // add mark and remove it from the boundary
    this.data[[x, y]] = {};
    delete this.bound[[x, y]];

    // add all non-marked surrounding points to boundary
    this._drawBound(x, y);

    if(line) {
      this._drawLine.apply(this, line);
    }

    return this;
  },

  unmark: function(x, y, line) {
    // first: clear line
    if(line) {
      this._clearLine.apply(this, line);
    }

    // remove mark
    delete this.data[[x, y]];

    // correct boundary status of this and all surrounding points
    for(var bx = x-1; bx <= x+1; bx++) {
      for(var by = y-1; by <= y+1; by++) {
        this._correctBound(bx, by);
      }
    }

    return this;
  },

  markSafe: function(x, y, line) {
    var valid = this.validMoves();
    for(var i = 0; i < valid.length; i++) {
      if(valid[i].join(',') == [x, y, line].join(',')) {
        this.mark(x, y, line);
        return this;
      }
    }
    throw "Invalid move: "+x+","+y+" ("+line+")";
  },

  markAll: function(marks) {
    for(var i = 0; i < marks.length; i++) {
      this.mark.apply(this, marks[i]);
    }
    return this;
  },

  validMoves: function() {
    var moves = [];
    for(var b in this.bound) {
      var bx = this.bound[b][0], by = this.bound[b][1];
      this.lines(bx, by).forEach(function(line) {
        moves.push([bx, by, line]);
      });
    }
    return moves;
  },

  lines: function(x, y) {
    var game = this;
    var lines = [];

    Morpion.ORIENTATIONS.forEach(function(d) {
      var dx = d[0], dy = d[1];

      // adj = [ahead, behind]
      var adj = [-1, 1].map(function(dir) {
        var nAdj = 0, cont = true, p;

        for(var i = 1; i <= game.len+game.dis; i++) {
          var p = [x+i*dx*dir, y+i*dy*dir];
          if(p in game.data) {
            if(cont) {
              nAdj++;
            }
            if(d in game.data[p]) {
              nAdj = Math.min(i-game.dis, nAdj);
              break;
            }
          } else {
            cont = false;
          }
        }

        return Math.min(game.len, nAdj);
      });

      for(var i = -adj[0]; i <= adj[1]-game.len; i++) {
        lines.push([x+i*dx, y+i*dy, dx, dy]);
      }
    });

    return lines;
  },

  isPlayed: function(x, y) {
    return [x, y] in this.data;
  },

  _drawBound: function(x, y) {
    for(var bx = x-1; bx <= x+1; bx++) {
      for(var by = y-1; by <= y+1; by++) {
        if(bx == x && by == y) {
          continue;
        } else if(!([bx, by] in this.data)) {
          this.bound[[bx, by]] = [bx, by];
        }
      }
    }
  },

  _correctBound: function(x, y) {
    if(!([x, y] in this.data)) {
      for(var bx = x-1; bx <= x+1; bx++) {
        for(var by = y-1; by <= y+1; by++) {
          if([bx, by] in this.data) {
            this.bound[[x, y]] = [x, y];
            return;
          }
        }
      }
    }
    delete this.bound[[x, y]];
  },

  _drawLine: function(x, y, dx, dy) {
    for(var i = 0; i <= this.len; i++) {
      var xi = x+i*dx, yi = y+i*dy;
      if(!([dx, dy] in this.data[[xi, yi]])) {
        this.data[[xi, yi]][[dx, dy]] = [];
      }
      this.data[[xi, yi]][[dx, dy]].push([x, y]);
    }
  },

  // clears the most recently added line (not any arbitrary line!)
  _clearLine: function(x, y, dx, dy) {
    for(var i = 0; i <= this.len; i++) {
      var xi = x+i*dx, yi = y+i*dy;
      var lines = this.data[[xi, yi]][[dx, dy]];
      for(var l = 0; l < lines.length; l++) {
        if(lines[l][0] == x && lines[l][1] == y) {
          lines.splice(l, 1);
          break;
        }
      }
      if(lines.length == 0) {
        delete this.data[[xi, yi]][[dx, dy]];
      }
    }
  }

};